---
title: "Clone ID with cardelino"
author:
- name: "Davis McCarthy"
  affiliation: 
  - EMBL-EBI, Hinxton, UK
  - St Vincent's Institute of Medical Research, Melbourne, Australia
- name: "Yuanhua Huang"
  affiliation:
  - EMBL-EBI, Hinxton, UK
package: cardelino
output:
    BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Clone ID with cardelino}
  %\VignetteEngine{knitr::rmarkdown}
  %VignetteEncoding{UTF-8}
---


```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE}
## To render an HTML version that works nicely with github and web pages, do:
## rmarkdown::render("vignettes/vignette.Rmd", "all")
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png',
    warning = FALSE, error = FALSE, message = FALSE)
library(ggplot2)
theme_set(theme_bw(12))
```

# Introduction

This document gives an introduction to and overview of inferring the clonal identity of cells using the `r Biocpkg("cardelino")` package using a given clonal structure.

`r Biocpkg("cardelino")` contains general tools for inferring the clonal or donor identity of cells from single-cell transcriptomic data, focusing on RNA-seq data. 
Specifically, the package features:

* Methods to infer the clonal identity of single cells from scRNA-seq read 
information, given a clonal structure;
* Methods to infer the donor identity of single cells from scRNA-seq read 
information, when donor genotype data is available for a set of possible donors.

Here, we focus on demonstrating the use of `cardelino` to probabilistically 
assign cells to clones when a clonal tree is provided.

# About the model

`r Biocpkg("cardelino")` can use variant information extracted from single-cell 
RNA-seq reads to probabilistically assign single-cell transcriptomes to 
individual clones.

Briefly, `r Biocpkg("cardelino")` is based on a Bayesian mixture model with a 
beta-binomial error model to account for sequencing errors as well as a 
gene-specific model for allelic imbalance between haplotypes and associated bias
in variant detection. Bayesian inference allows the model to account for 
uncertainty in model parameters and cell assignments.

We assume that clones are tagged by somatic mutations, and that these mutations
are known (e.g. from exome sequencing or equivalent). Given a set of known 
mutations, these sites can be interrogated in scRNA-seq reads to obtain evidence 
for the presence or absence of each mutation in each cell. As input, the model 
requires the count of reads supporting the alternative (mutant) allele at each 
mutation site, the total number of reads overlapping the mutation site 
("coverage"). 

Typically, coverage of somatic mutations in scRNA-seq data is very
sparse (most mutation sites in a given cell have no read coverage), but the 
`r Biocpkg("cardelino")` model accounts for this sparsity and aggregates 
information across all available mutation sites to infer clonal identity.

# Clone ID with a clonal tree provided

In many clone ID scenarios, a clonal tree is known. That is, we have been able 
to infer the clones present in the sampled cell population, for example using 
bulk or single-cell DNA-seq data, and we know which mutations are expected to be
present in which clones.

To infer the clonal identity of cells when a clonal tree is provided, 
`r Biocpkg("cardelino")` requires the following input data:

* A: a variant x cell matrix of integer counts, providing the number of reads
supporting the alternative allele for each variant in each cell;
* D: a variant x cell matrix of integer counts, providing the total number of 
reads overlapping each variant in each cell;
* C: a variant x clone "configuration" matrix of binary values providing the 
clone-variant configuration by indicating which mutations are expected to be 
present in which clones.

The configuration matrix, C, can be provided by other tools used to infer the 
clonal structure of the cell population. For example, the package 
`r CRANpkg("Canopy")` can be used to infer a clonal tree from DNA-seq data and 
the "Z" element of its output is the configuration matrix.

Here, we demonstrate the use of `r Biocpkg("cardelino")` to assign 428 cells to 
clones identified with `r CRANpkg("Canopy")` using 34 somatic mutations. 

We load the package and the example clone ID dataset distributed with the 
package.

```{r load-pkg}
library(cardelino)
data(example_donor)
```

The clonal tree inferred by Canopy for this donor consists of four clones, 
including a "base" clone ("clone1") that has no sublconal somatic mutations 
present.

```{r plot-tre}
plot_tree(tree, orient = "v")
```

The included dataset contains the A and D matrices, so combined with the Canopy 
tree object provided, we have the necessary input to probabilistically assign 
cells to clones.

```{r cell-assign}
assignments <- clone_id(A, D, C = tree$Z)
names(assignments)
```

We can visualise the cell-clone assignment probabilities as a heatmap.

```{r prob-heatmap}
prob_heatmap(assignments$prob)
```

We recommend assigning a cell to the highest-probability clone if the highest 
posterior probability is greater than 0.5 and leaving cells "unassigned" if they
do not reach this threshold. The `assign_cells_to_clones` function conveniently
assigns cells to clones based on a threshold and returns a data.frame with the
cell-clone assignments.

```{r assign-cell-clone-easy}
df <- assign_cells_to_clones(assignments$prob)
head(df)
table(df$clone)
```

# Reading in data from VCF files

In the genomics field, genotype data is most commonly stored in VCF 
([variant call format](https://github.com/samtools/hts-specs)) files. There are 
many possible ways to extract the data required by `cardelino` from a VCF file,
here we show just one approach using convenience functions in `cardelino`:

Read in data from VCF included in the package. 

```{r read-vcf-data}
vcf <- read_vcf(system.file("extdata", "cell_example.mpileup.vcf.gz", 
                   package = "cardelino"))
input_data <- get_snp_matrices(vcf)
```

Read in Canopy tree results for the same individual.

```{r read-canopy-data}
canopy <- readRDS(system.file("extdata", "canopy_results.example.rds", 
                   package = "cardelino"))
C <- canopy$tree$Z
```

Be careful to ensure that the same variant IDs are used in both data sources.

```{r correct-variant-ids}
rownames(C) <- gsub("chr", "", gsub(":", "_", gsub("_.*", "", rownames(C))))
```

Now we can run the clone ID function to assign cells to clones.

```{r run-cell-assign}
assignments <- clone_id(input_data$A, input_data$D, C)
```

As above, we can easily assign cells to clones using a threshold on the
posterior probabilities output by the `clone_id` function.

```{r assign-cell-clone-vcf}
df <- assign_cells_to_clones(assignments$prob)
table(df$clone)
```


# Session information {.unnumbered}

```{r}
sessionInfo()
```
