---
title: "Donor ID with cardelino"
author:
- name: "Davis McCarthy"
  affiliation: 
  - EMBL-EBI, Hinxton, UK
  - St Vincent's Institute of Medical Research, Melbourne, Australia
- name: "Yuanhua Huang"
  affiliation:
  - EMBL-EBI, Hinxton, UK
package: cardelino
output:
    BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Clone ID with cardelino}
  %\VignetteEngine{knitr::rmarkdown}
  %VignetteEncoding{UTF-8}
---


```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE}
## To render an HTML version that works nicely with github and web pages, do:
## rmarkdown::render("vignettes/vignette.Rmd", "all")
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png',
    warning = FALSE, error = FALSE, message = FALSE)
library(ggplot2)
library(BiocStyle)
theme_set(theme_bw(12))
```

# Introduction

This document gives an introduction to and overview of inferring the **donor** 
identity of cells using the `r Biocpkg("cardelino")` package. Many single-cell
genomics studies can be enhanced by multiplexing donors, that is cells from 
multiple genetically distinct individuals ("donors") are assayed together in a 
mixed population. Such experimental designs can be powerful for reducing cost
of studies assaying cells from many donors and for enabling robustness to 
inevitable batch effects. 

When cells from multiple donors are mixed together, the donor identity of each
cell is unknown at the time of sequencing. However, natural genetic variation
(specifically, single nucelotide polymorphisms) act as natural barcodes 
capturing the donor identity of each cell. Appropriate computational methods, 
such as presented here, can infer the donor identity for each cells and thus
"demultiplex" a population of cells from multiple donors in preparation of the
dataset for further downstream analysis.

`r Biocpkg("cardelino")` contains general tools for inferring the clonal or 
donor identity of cells from single-cell transcriptomic data, focusing on 
RNA-seq data. Specifically, the package features:

* Methods to infer the **clonal identity** of single cells from scRNA-seq read 
information, given a clonal structure;
* Methods to infer the **donor identity** of single cells from scRNA-seq read 
information, when donor genotype data is available for a set of possible donors.

Here, we focus on demonstrating the use of `cardelino` to probabilistically 
assign cells to donors when donor genotypes are provided.

# About the model

`r Biocpkg("cardelino")` can use variant information extracted from single-cell 
RNA-seq reads to probabilistically assign single-cell transcriptomes to 
specific donor individuals.

Briefly, `r Biocpkg("cardelino")` is based on a Bayesian mixture model with a 
beta-binomial error model to account for sequencing errors as well as a 
gene-specific model for allelic imbalance between haplotypes and associated bias
in variant detection. Bayesian inference allows the model to account for 
uncertainty in model parameters and cell assignments.

We assume that donors are uniquely tagged by single nucleotide polymorphisms 
(SNPs), and that these mutations are known (e.g. chip genotyping or equivalent).
Given a set of known SNPs, these sites can be interrogated in scRNA-seq reads to
obtain evidence for the genotypes of each SNP in each cell. As input, the model 
requires the count of reads supporting the alternative allele at each variant
site and the total number of reads overlapping the mutation site ("coverage") in
each cell, and the known genotypes of the donors at those SNPs (encoded as 
counts of the alternative allele, i.e. 0, 1, or 2).

Typically, coverage of SNPs in scRNA-seq data is very sparse (most SNPs in a 
given cell have no read coverage), but the `r Biocpkg("cardelino")` model 
accounts for this sparsity and aggregates information across all available 
variant sites to infer donor identity. The model returns a posterior probability
of assignment to each donor and also the probability that the "single cell" is 
in fact a "doublet" (two, or more, cells merged together in the assay).

# Donor ID with donor genotypes provided

In many donor ID scenarios, reference genotype information will be available 
for each donor. That is, donors have been genotyped using a SNP chip platform, 
exome sequencing or similar, so we know the expected genotype of each SNP for
each of the multiplexed donors from which our cells come.

To infer the donor identity of cells when a donor genotypes are provided, 
`r Biocpkg("cardelino")` requires the following input data:

* **cell VCF file:** a variant call format (VCF) file from which we can extract 
the variant x cell matrices of integer counts of the number of reads supporting
the alternative allele for each variant in each cell and the total number of 
reads overlapping each variant in each cell;
* **donor VCF file:** a VCF file containing genotypes for each of the multiplexed
donors.

For compatibility with genomics pipelines, we accept variant data in VCF 
([variant call format](https://github.com/samtools/hts-specs)) files. Both VCF
files should be indexed.

As the `cardelino` model uses the read count data from the cell VCF file, it 
does not require "true genotyping" for the SNPs in the cells. We have had 
success using the `mpileup` and `call` methods in the 
[`bcftools`](http://www.htslib.org/doc/bcftools.html) software to generate cell 
VCF files. However, any appropriate method that produces a valid VCF and 
contains correct "AD" and "DP" fields in the VCF will work fine. Both VCF files
are read into memory, so we recommend filtering large VCF files before applying 
`cardelino`. We suggest filtering the cell VCF file to just keep variants with 
read coverage in at least a small percentage of cells (say 3%) and then to 
filter both the cell and donor VCFs to contain just the variants that appear in
both VCF files. Again, `bcftools` has good filtering methods for these purposes.

Here, we demonstrate the use of `r Biocpkg("cardelino")` to assign 428 cells to 
clones identified with `r CRANpkg("Canopy")` using 34 somatic mutations. 

We load the package and look at an example donor ID dataset distributed with the 
package.

```{r load-pkg}
library(cardelino)
cell_vcf <- read_vcf(system.file("extdata", "cells.donorid.vcf.gz", 
                                 package = "cardelino"))
```

In this particular example, cells from three donors were mixed together. If we
run probabilistic principal component analysis (PPCA) on the raw "genotypes" of 
the cells we can get a feel for the grouping of cells by donor.

```{r ppca-raw, fig.height=4, fig.width=6}
raw_geno <-  VariantAnnotation::genotypeToSnpMatrix(
    VariantAnnotation::geno(cell_vcf, "GT"),
    ref = VariantAnnotation::ref(cell_vcf),
    alt = VariantAnnotation::alt(cell_vcf))
pp <- pcaMethods::ppca(as(raw_geno$genotypes, "numeric"))
df <- data.frame(PPCA1 <- pp@scores[, 1], PPCA2 <- pp@scores[, 2])
ggplot(df, aes(PPCA1, PPCA2)) +
        geom_point(alpha = 0.5) +
        theme_bw()
```

To infer donor identity, we use the `donor_id` function in 
`r Biocpkg("cardelino")`. VCFs can be supplied either as file paths or as
`CollapsedVCF` objects as produced when reading in data with the `read_vcf`
function as above. Both usages are demonstrated below.

The example donor VCF file here contains genotype information for three donors 
that contributed cells to this multiplexed cell population. We utilise the 
option to detect doublets by setting `check_doublet = TRUE`. The `donor_id` 
function returns a list of outputs, one of which is the `assigned` data.frame
with inferred donor ID (either a donor ID, "doublet" or "unassigned").

```{r plot-tre}
ids <- donor_id(cell_vcf,
                system.file("extdata", "donors.donorid.vcf.gz", 
                                 package = "cardelino"),
                 check_doublet = TRUE)
table(ids$assigned$donor_id)
```

The `assigned` data.frame contains the key information from the donor 
assignment, including the highest posterior probability value across the donors
(`prob_max`), the posterior doublet probability (`prob_doublet`) and the 
number of variants with non-zero read coverage for the cell (`n_vars`). By default, if `n_vars` is less than 10, we expect that donor ID is unlikely to be
accurate and return donor ID as "unassigned" for these cells. This threshold 
can be adjusted with the `n_vars_threshold` argument.

```{r head-assigned}
head(ids$assigned)
```

We can see that doublets tend to be inferred for "cells" with fewer variants 
with coverage (that is for cells with less donor information). In this case, 
reassuringly, the inferred doublets include known "minibulk" wells (10 cells 
from multiple donors processed as one "cell") and positive control wells 
(details not shown).

```{r plot-doublet, fig.height=4, fig.width=6}
ggplot(ids$assigned, aes(n_vars, prob_doublet, colour = donor_id)) +
    geom_point(alpha = 0.5) +
    theme_bw()
```

For most cells, the `cardelino` model is very confident about donor assignment
(posterior probability of 1 for best-matching donor). Once there are more than
about 30 variants with read coverage we can assign cells to donors with very 
high confidence.

```{r plot-postprob, fig.height=4, fig.width=6}
ggplot(ids$assigned, aes(n_vars, prob_max, colour = donor_id)) +
    geom_point(alpha = 0.5) +
    scale_x_log10() +
    theme_bw()
```

When we map our donor ID back onto the PPCA plot we see that the clusters 
evident originally indeed correspond to the individual donors. Here, PPCA 
component 1 clearly separates the donors, while PPCA component 2 appears to
reflect coverage of variants (unassigned cells with low coverage appear at low
values in PPCA2).

```{r plot-ppca-donor, fig.height=4, fig.width=6}
df$donor_id <- ids$assigned$donor_id
ggplot(df, aes(PPCA1, PPCA2, colour = donor_id)) +
        geom_point(alpha = 0.5) +
        theme_bw()
```

With donor identity now inferred for these cells, the donor identities can be
linked back to the expression profiles for the cells for downstream analysis.

# Coping with "extreme" multiplexing

The high sensitivity and specifity of donor demultiplexing from scRNA-seq data
enables the possibility of "extreme" multiplexing: studies assaying cells from 
hundreds of donors, with dozens of donors multiplexed per assay (say "per run").

In such cases, it might be convenient to have a single donor reference genotype
VCF file with hundreds of donor genotypes. The approach shown above will work 
when given a donor genotype VCF with hundreds of donors, but if only a subset of
donors are multiplexed in a given run, then the default approach will be slow 
and likely call too many doublets (as we form doublet genotype profiles from 
all pairs of donor genotypes, so this scales as the square of the number of 
donors; with a large number of donors there is a *very* large number of possible
doublet genotypes to choose from).

To handle these situations, we propose the following two-step approach:

* **First pass:** run `donor_id` with the full donor genotype VCF, without doublet
detection;
* **Second pass:** run `donor_id` with doublet detection, but specifying the 
function to use just the donors that had at least once cell assigned in the 
first pass.

We demonstrate this approach on the same dataset as used above (although there 
are only three donors in this case, this same set of commands will work for 
hundreds of donors).

**First pass**: run without doublet detection.

```{r extreme-first}
donor_vcf <- read_vcf(system.file("extdata", "donors.donorid.vcf.gz", 
                                 package = "cardelino"))
ids_sing <- donor_id(cell_vcf, donor_vcf, check_doublet = FALSE)
```

**Second pass**: run with doublet detection, but just on the donors identified
in the first pass.

```{r extreme-second, eval=TRUE}
ids_doub <- donor_id(cell_vcf, donor_vcf, check_doublet = TRUE,
                        donors = unique(ids_sing$assigned$donor_id))
table(ids_doub$assigned$donor_id)
```

In this case, the results are exactly the same as with the default approach 
demonstrated above (as they should be), but the flexibility of this approach for
dealing with large numbers of donor genotypes should be clear.

# Session information {.unnumbered}

```{r}
sessionInfo()
```
