---
title: "Expression confounding in cell clonal assignment"
author: "Yuanhua Huang"
to_working_dir: "/Users/huangh/research/cellseq/clonality/fibroblast/reports/"
output:
  html_document:
        toc: true
        toc_float: true
        theme: journal
        highlight: pygments
        number_sections: true
        code_folding: hide
---

This notebook gives a test analysis of cell assignemnts to clones with mixture 
models: two-component Bernoulli or binomial model. Here, the parameters are the 
mean success rate of both models. The estimation and the cell assignments are 
highly coupled. Therefore, we use two standard methods to infer this mixture 
model: EM algorithm and Gibbs sampling.

EM algorithm is much more efficient than Gibbs sampling, but EM algorithm only 
gives a point estimate of these parameters. However, a single fixed parameter 
value may fail to capture the variability of these parameters for the alteration 
reads distribution across variants and cells. Therefore, the cell assignments 
may overfitted to a single parameter value.

Alternatively, we could use the Bayesian inference, which marginalize these 
parameters when assigning cells to clones and may give more robust inference of 
the cell assignments.

**See more**: 
[Details on methods](https://drive.google.com/a/ebi.ac.uk/file/d/1Aclmfk23v0wM4lxrtvHWmzpMIWDvLIoa/view), 
[R Codes for EM algorithm](https://github.com/davismcc/cardelino/blob/master/R/cell_assign_EM.R), and 
[R Codes for Gibbs sampling](https://github.com/davismcc/cardelino/blob/master/R/cell_assign_Gibbs.R)


# Load full genotyp data
```{r}
setwd("/Users/huangh/research/cellseq/clonality/fibroblast")
fig_dir="/Users/huangh/research/cellseq/clonality/figures/technical/"

library(vcfR)

system.time(
    vcf <- read.vcfR("genotypes/cells_merged_clone_id.vcf.gz")
)
dp_full <- extract.gt(vcf, element = "DP", as.numeric = TRUE)
ad_full <- extract.gt(vcf, element = "AD")
ad_full <- masplit(ad_full, record = 2, sort = FALSE)
rownames(ad_full) <- rownames(dp_full) <- paste0(vcf@fix[, 1], "_", vcf@fix[, 2])
```

```{r}
donor <- "zoxy"
donor_id <- "zoxy"
#"rozh", "vabj", "xugn"#"qolg" #"lexy" #"vass" #"zoxy" "wuye"
```

# Load donor data
```{r}
setwd("/Users/huangh/research/cellseq/clonality/fibroblast")

library(vcfR)

## Load data
#donor <- "zoxy" #"lexy" #"vass" #"zoxy" "wuye"
canopy <- readRDS(paste0("canopy/canopy_results.", donor, ".rds"))
sce <- readRDS(paste0("sces/sce_", donor, "_qc.rds"))

## Match exome-seq and scRNA-seq data
# match cells
mm <- match(colnames(sce), colnames(dp_full))
mm <- mm[!is.na(mm)]

# match variants
C_row_names <- row.names(canopy$tree$Z)
for (i in seq_len(length(C_row_names))){
  temp_name <- strsplit(strsplit(C_row_names[i], split="_")[[1]][1], split=":")[[1]]
  C_row_names[i] <- paste0("chr", temp_name[1], "_", temp_name[2])
}

dp <- dp_full[rownames(dp_full) %in% C_row_names,][,mm]
ad <- ad_full[rownames(ad_full) %in% C_row_names,][,mm]
idx_var <- match(rownames(dp), C_row_names)
idx_var <- idx_var[!is.na(idx_var)]

# filter variants with no cells
idx_var_use <- rowMeans(is.na(dp))<1

A <- ad[idx_var_use,]
D <- dp[idx_var_use,]

mapped.tree <- canopy$tree
mapped.tree$Z <- mapped.tree$Z[idx_var[idx_var_use],]
mapped.tree$sna <- mapped.tree$sna[idx_var[idx_var_use],]

Psi <- mapped.tree$P[, 1]
Config <- mapped.tree$Z

table(canopy$tree$Z %*% 2**seq_len(dim(Config)[2]))
table(mapped.tree$Z %*% 2**seq_len(dim(Config)[2]))
```

## Filtering cells with limited variants
```{r}
library(ggpubr)
nvars_genotyped <- colSums(!is.na(A))
print(sum(nvars_genotyped>0.5))

data.frame(nvars_genotyped = colSums(!is.na(A))) %>%
    ggdensity(x = "nvars_genotyped",
              add = "median", rug = TRUE,
              color = "#00AFBB", fill = "#00AFBB")

A <- A[, nvars_genotyped > 0.5]
D <- D[, nvars_genotyped > 0.5]
sce <- sce[, nvars_genotyped > 0.5]
```

# EM algorithm for cell assignment
## Bernoulli base model
```{r}
## Cell assignment with EM Algorithm
# source("/Users/huangh/Dropbox/mycodes/cellseq/clonality/models/cell_assign_EM.R")
# devtools::install_github("davismcc/cardelino")
library(cardelino)

Config <- mapped.tree$Z
Psi <- mapped.tree$P[, 1]

# Config[,3] <- as.integer((Config[,3]+Config[,4])>0)
# Config <- Config[,1:3]

vass_Bern_EM <- cell_assign_EM(A, D, Config, Psi=rep(1/ncol(Config), ncol(Config)), model="Bernoulli")
prob_Bern_EM <- vass_Bern_EM$prob
vass_Bern_EM$theta #false positive rate; 1-false negative rate
vass_Bern_EM$logLik

conf_labels <- rep(0, dim(prob_Bern_EM)[1])
prob_gaps <- rep(0, dim(prob_Bern_EM)[1])
for (i in 1:dim(prob_Bern_EM)[1]){
  prob_sorted <- sort(prob_Bern_EM[i,], decreasing=T)
  prob_gaps[i] <- prob_sorted[1] - prob_sorted[2]
  if (prob_gaps[i] > 0.2){
    conf_labels[i] = which.max(prob_Bern_EM[i,])
  }
}
clone_ids <- colnames(prob_Bern_EM)
clone_perc <- clone_ids
for (i in seq_len(length(clone_perc))){
  conf_frac <- sum(conf_labels == i) / length(conf_labels)
  clone_perc[i] <- paste0(clone_perc[i], ": ", round(conf_frac*100,digits=1), "%")
}
colnames(prob_Bern_EM) <- clone_perc
```

## Heatmap with Canopy tree
```{r}
library(gridExtra)
library(ggplot2)
library(reshape2)
source("/Users/huangh/Dropbox/mycodes/cellseq/clonality/models/plot_tree.R")

snv_label <- mut.label(mapped.tree)

fig_tree <- ggtree.canopy(mapped.tree, orient = "v") + 
  ggtitle(paste("Cell assignment to clones:", donor)) + 
  xlab("Clonal tree from bulk exome-seq") + pub.theme()

hc <- hclust(dist(prob_Bern_EM))
nba.m <- melt(prob_Bern_EM[hc$order,])
fig_assign <- ggplot(nba.m, aes(Var2, Var1, fill = value)) + 
  geom_tile(show.legend=T) +
  scale_fill_gradient(low = "white", high = "firebrick4") +
  ylab(paste("clonal asignment probability for ", length(hc$order), "single cells")) + 
  heatmap.theme() + pub.theme()

fig_both <- grid.arrange(fig_tree, fig_assign, nrow=2, heights=c(0.35, 0.63))
fig_file <- paste0(fig_dir, "/cell_assign_proba_", donor, ".png")
ggsave(fig_file, plot=fig_both, device = NULL, path = NULL,
  scale = 1, width = 7, height = 8, units = "in", dpi = 600)
fig_file
```


# Technical effects
## PCA on variant coverages
```{r}
conf_labels_use <- as.factor(conf_labels)
conf_labels_use[conf_labels==0] <- "U"

D[is.na(D)] <- 0
df_f <- t(D[rowSums(D)>0,])

ir.pca <- prcomp(df_f)
var_frac <- ir.pca$sdev**2/sum(ir.pca$sdev**2)
dat <- data.frame(PC1=ir.pca$x[,1], PC2=ir.pca$x[,2], 
                  total_counts=sce$total_counts, Clone=conf_labels_use)

fig_pca <- ggplot(dat, aes(x=PC1, y=PC2, color=Clone)) + geom_point(size=2)+
  labs(title=paste("PCA on dp of", dim(df_f)[2], "SNVs:", donor),
       x = paste0("PC 1: ", round(var_frac[1]*100, digits=1), "% var"),
       y = paste0("PC 2: ", round(var_frac[2]*100, digits=1), "% var"))
fig_pca <- fig_pca + theme_classic() + pub.theme(11)

corr_coef <- round(cor(dat$PC1, dat$total_counts), digits = 2)
fig_p1 <- ggplot(dat, aes(x=PC1, y=total_counts, color=Clone)) + 
  geom_point(size=2)+
  labs(title=paste("Seq depth Corr. R =", corr_coef, ":", donor), 
       x = "PC 1", y = "Total reads count")
fig_p1 <- fig_p1 + scale_y_log10() + theme_classic() + pub.theme(11)

fig_pca_both <- grid.arrange(fig_pca, fig_p1, ncol=2)
ggsave(paste0(fig_dir, "/PCA_dp_", donor, ".png"),
       plot=fig_pca_both, height = 3.5, width = 9, units="in", dpi=600)
fig_pca_both
```

## Assignment probability and total features
```{r}
#library(scales)
library(ggplot2)
prob_var <- rowSums((prob_Bern_EM - rowMeans(prob_Bern_EM))**2)
# for (i in seq_len(length(prob_var))){
#   prob_var[i] <- entropy::entropy(prob_Bern_EM[i,])
# }
prob_dat <- data.frame(prob_Bern_EM, prob_var)
colnames(prob_dat) <- c(colnames(Config), "diversity")

for (i in seq_len(dim(prob_dat)[2])){
  colnames(prob_dat)[i] <- paste0(colnames(prob_dat)[i], ": R=",
                                 round(cor(prob_dat[,i], sce$total_features), 
                                      digits = 2))
}

nba.m <- melt(prob_dat)
prob_melt <- data.frame(nba.m, total_features=rep(sce$total_features, 
                                                  dim(prob_dat)[2]))

fig_prob <- ggplot(prob_melt, aes(x=value, y=total_features)) + 
  geom_point() + facet_wrap(~variable) +
  labs(title=paste("total features and cell assignment:", donor),
       x = "clonal assignment probability", y = "Total features")
fig_prob <- fig_prob + 
  scale_y_log10(breaks = scales::trans_breaks("log10", function(x) 10^x),
                labels = scales::trans_format("log10", scales::math_format(10^.x))) +
  #scale_y_continuous(labels = NotFancy) +#label=scientific_format()
  theme_classic() + pub.theme(11)
ggsave(paste0(fig_dir, "/proba_total_features_", donor, ".png"),
       plot=fig_prob, height = 6, width = 10, units="in", dpi=600)
fig_prob
```

### Assigned clone and cell condition 
```{r}
conf_labels_use[is.na(conf_labels_use)] = 0
cond_dat <- data.frame(condition=sce$well_condition, Clone=conf_labels_use)
cond_melt <- melt(table(cond_dat))
cond_melt$Clone[cond_melt$Clone==0] = "U"

fig_cond <- ggplot(cond_melt, aes(condition, Clone, fill = value)) + 
  geom_tile(show.legend=F) +
  geom_text(aes(label = round(value, 1))) +
  scale_fill_gradient(low = "white", high = "firebrick4") +
  labs(title="cell clonal assignment and conditions", x="", y="Assigned Clones") + 
  theme_gray() + pub.theme(size=11) + 
  theme(axis.text.x = element_text(angle=0),
        panel.grid.major = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        axis.ticks = element_blank())
ggsave(paste0(fig_dir, "/condition_clone_", donor, ".png"),
       plot=fig_cond, height = 4, width = 5, units="in", dpi=600)
fig_cond
```

