#' Sythetic reads generator for genetic variants
#' 
#' There are following steps to generate the simulated reads counts for variants
#' in single cells:
#' 1) given the clonal genotype and the clonal prevalence, the genotypes of 
#' cells will be generated following a multinormial distribution. 
#' 2) given the distribution of reads coverage, e.g., a set of read coverage 
#' from real data, the total reads of each variant will be generated by random 
#' sampling.
#' 3) the allelic frequence of each variant will be generated by following a 
#' beta distribution.
#' 4) Given the genotype of a cell, if the mutation exists in a cell, the 
#' alteration read counts will be generated by a binomial distribution, 
#' papramitized the allelic frequency, sampled from step 3.
#' 5) Given the genotype of a cell, if the mutation doesnot exist in a cell, 
#' the alteration read counts will be genenrated by a binomial distribution,
#' paramerized by the technical error rate.
#' 6) Given the missing rate, NA value will be given to mask the read counts
#' 
#' @export
#' 
#' @param D A vector of integers. Sequencing depth for variants
#' @param C A matrix of binary values. The clone-variant configuration, whcih 
#' encodes the phylogenetic tree structure. This is the output Z of Canopy
#' @param Psi A vector of float. The fractions of each clone, output P of Canopy
sim_read_count <- function(C, D, Psi=NULL, means=c(0.03, 0.45), 
                           vars=c(0.1, 0.1), missing_rate=NULL, cell_num=300){
  M <- cell_num   #number of cells
  K <- dim(C)[2]  #number of clones
  N <- dim(C)[1]  #number of variants
  
  shape1s <- means / vars
  shape2s <- 1.0 /vars - shape1s
  
  # genotype for cells H_sim, and clone labels I_sim
  if(is.null(Psi)){
    Psi <- rep(1/K, K)
  }
  I_sim <- rmultinom(M, 1, prob=Psi)
  H_sim <- C %*% I_sim
  
  # generate p, D and A
  p_sim <- matrix(0, N, 2)
  A_sim <- matrix(NA, N, M)
  D_sim <- matrix(NA, N, M)
  D_non_0 <- D[which(D>0)]
  for (i in seq_len(N)){
    p_sim[i,1] <- rbeta(1, shape1s[1], shape2s[1])
    p_sim[i,2] <- rbeta(1, shape1s[2], shape2s[2])
    for(j in seq_len(M)){
      if(is.null(missing_rate)){
        D_sim[i,j] = sample(D[i,], 1, replace = TRUE)
      } else{
        is_mising <- rbinom(1, 1, missing_rate)
        if (!is_mising){
          # D_sim[i,j] = sample(D_non_0, 1, replace = TRUE) 
          # variant specific expression
          D_sim[i,j] = sample(D[i,which(D[i,]>0)], 1, replace = TRUE) 
        }
      }
      if(!is.na(D_sim[i,j])){
        A_sim[i,j] = rbinom(1, D_sim[i,j], p_sim[i,H_sim[i,j]+1])
      }
    }
  }
  return_list <- list("A_sim"=A_sim, "D_sim"=D_sim, "I_sim"=t(I_sim), 
                      "H_sim"=H_sim, "p_sim"=p_sim)
  return_list
}


