
#' Sythetic reads generator for genetic variants
#' 
#' There are following steps to generate the simulated reads counts for variants
#' in single cells:
#' 1) given the clonal genotype and the clonal prevalence, the genotypes (i.e, 
#' the clone) of cells will be generated following a multinormial distribution. 
#' 2) given the distribution of reads coverage, e.g., a matrix of read coverage 
#' from real data, (variant specific), the total reads of each variant will be 
#' generated by random sampling. Note, the missing rate is gorvened by this 
#' matrix.
#' 3) the allelic frequence of each variant will be generated by following a 
#' beta distribution with paramters of mean and varince.
#' 4) Given the genotype of a cell, if the mutation exists in a cell, the 
#' alteration read counts will be generated by a binomial distribution, 
#' papramitized the allelic frequency, sampled from step 3.
#' 5) Given the genotype of a cell, if the mutation does not exist in a cell, 
#' the alteration read counts will be genenrated by a binomial distribution,
#' paramerized by the technical error rate.
#' 
#' @export
#' 
#' @param C A matrix of binary values. The clone-variant configuration, whcih 
#' encodes the phylogenetic tree structure, and the genotype of each clone.
#' @param D A matrix of integers. Sequencing depth for N variants across x 
#' cells (ideally >100 cells). NA means 0 here.
#' @param Psi A vector of float. The fractions of each clone. If NULL, set a 
#' uniform distribution.
#' @param means A vector of two floats. The mean theta_1 (false positive rate) 
#' and the mean theta_2 (true positive rate).
#' @param vars A vector of two floats. The variance of theta_1 and theta_2.
#' @param cell_num A interger. The number of cells to generate.
#' @param permute_D A boolean value. If True permute variants in D.
#' 
#' @return a list containing \code{A_sim}, a matrix for alteration reads, 
#' \code{A_sim}, a matrix for total reads, \code{I_sim}, a matrix for clonal 
#' label, \code{H_sim}, a matrix for genotype, and \code{H_sim}, a matrix for 
#' theta_1 and theta_2 in each varaint.
#' 
sim_read_count <- function(C, D, Psi=NULL, means=c(0.03, 0.45), vars=c(0.1,0.1), 
                           cell_num=300, permute_D=FALSE){
  M <- cell_num   #number of cells
  K <- dim(C)[2]  #number of clones
  N <- dim(C)[1]  #number of variants
  
  shape1s <- means / vars
  shape2s <- 1.0 / vars - shape1s
  
  if (permute_D == TRUE){
    D <- D[sample(nrow(D)), ]
  }
  
  # genotype for cells H_sim, and clone labels I_sim
  if(is.null(Psi)){
    Psi <- rep(1/K, K)
  }
  I_sim <- rmultinom(M, 1, prob=Psi) # Clonal label for each cell
  H_sim <- C %*% I_sim               # Genotype for each cell
  
  # generate p, D and A
  p_sim <- matrix(0, N, 2)   # Theta_1 and theta_2 for each variants
  A_sim <- matrix(NA, N, M)  # Alteration counts matrxi
  D_sim <- matrix(NA, N, M)  # Total counts matrix
  for (i in seq_len(N)){
    p_sim[i,1] <- rbeta(1, shape1s[1], shape2s[1])
    p_sim[i,2] <- rbeta(1, shape1s[2], shape2s[2])
    for(j in seq_len(M)){
      D_sim[i,j] = sample(D[i,], 1, replace = TRUE)
      if(!is.na(D_sim[i,j])){
        A_sim[i,j] = rbinom(1, D_sim[i,j], p_sim[i,H_sim[i,j]+1])
      }
    }
  }
  return_list <- list("A_sim"=A_sim, "D_sim"=D_sim, "I_sim"=t(I_sim), 
                      "H_sim"=H_sim, "p_sim"=p_sim)
  return_list
}


